File: 139.单词拆分.c
/*
 * @lc app=leetcode.cn id=139 lang=c
 *
 * [139] 单词拆分
 */

// @lc code=start

bool wordBreak(char *s, char **wordDict, int wordDictSize)
{
  int i, j, k, *dp;
  dp = (int *)malloc(sizeof(int) * (strlen(s) + 1));
  dp[0] = 1;

  for (i = 1; i < strlen(s) + 1; i++)
  {
    for (j = 0; j <= i; j++)
    {
      char *str = (char *)malloc(sizeof(char) * (i - j + 1));
      for (k = j; k < i; k++)
        str[k - j] = s[k];
      str[k - j] = '\0';
      for (k = 0; k < wordDictSize; k++)
      {
        if (strcmp(wordDict[k], str) == 0 && dp[j] == 1)
        {
          dp[i] = 1;
          break;
        }
      }
    }
  }

  if (dp[strlen(s)] == 1)
    return true;
  else
    return false;
}
// @lc code=end

File: store/139.单词拆分.js
/*
 * @lc app=leetcode.cn id=139 lang=javascript
 *
 * [139] 单词拆分
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  let dp = new Array(s.length + 1).fill(false);
  dp[0] = true;

  for (let i = 1; i < s.length + 1; i++) {
    for (let j = 0; j < i; j++) {
      const str = s.slice(j, i);
      if (wordDict.indexOf(str) !== -1 && dp[j]) dp[i] = true;
    }
  }

  return dp[s.length];
};
// @lc code=end
